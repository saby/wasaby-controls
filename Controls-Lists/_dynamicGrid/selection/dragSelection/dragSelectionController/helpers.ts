import { TPoint } from './types';
import { TResizerDirection } from '../shared/types';

export function getDragStartParams(
    startRect: DOMRect,
    endRect: DOMRect,
    direction: TResizerDirection
) {
    let xOffset: number;
    let yOffset: number;
    let xMinOffset: number;
    let yMinOffset: number;

    const startPoint: TPoint = {
        x: undefined,
        y: undefined,
    };

    // Начальная точка это либо левый верхний угол, либо правый нижний.
    // Координата Y
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'topRight':
            startPoint.y = endRect.y + endRect.height;
            yOffset = startRect.y - startPoint.y;
            break;
        case 'bottom':
        case 'right':
        case 'bottomLeft':
        case 'bottomRight':
            startPoint.y = startRect.y;
            yOffset = endRect.y + endRect.height - startRect.y;
            break;
    }

    // Координата X
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            startPoint.x = endRect.x + endRect.width;
            xOffset = startRect.x - startPoint.x;
            break;
        case 'bottom':
        case 'right':
        case 'topRight':
        case 'bottomRight':
            startPoint.x = startRect.x;
            xOffset = endRect.x + endRect.width - startRect.x;
            break;
    }

    // Минимальное смещение по Y
    switch (direction) {
        case 'top':
        case 'topLeft':
        case 'topRight':
            yMinOffset = -1 * endRect.height;
            break;
        case 'bottom':
        case 'bottomLeft':
        case 'bottomRight':
            yMinOffset = startRect.height;
            break;
        case 'left':
        case 'right':
            yMinOffset = yOffset;
            break;
    }

    // Минимальное смещение по X
    switch (direction) {
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            xMinOffset = -1 * endRect.width;
            break;
        case 'right':
        case 'topRight':
        case 'bottomRight':
            xMinOffset = startRect.width;
            break;
        case 'top':
        case 'bottom':
            xMinOffset = xOffset;
            break;
    }

    return {
        startPoint,
        xOffset,
        yOffset,
        xMinOffset,
        yMinOffset,
    };
}

export function getNewOffset(
    cellRect: DOMRect,
    startPoint: TPoint,
    direction: TResizerDirection,
    minXOffset: number,
    minYOffset: number
): {
    x: number;
    y: number;
} {
    let newOffsetX: number;
    let newOffsetY: number;

    // Координата Y
    switch (direction) {
        case 'top':
        case 'topLeft':
        case 'topRight':
            newOffsetY = cellRect.y - startPoint.y;
            newOffsetY = newOffsetY > 0 ? minYOffset : newOffsetY;
            break;
        case 'bottom':
        case 'bottomLeft':
        case 'bottomRight':
            newOffsetY = cellRect.y + cellRect.height - startPoint.y;
            newOffsetY = newOffsetY < 0 ? minYOffset : newOffsetY;
            break;
        case 'left':
        case 'right':
            newOffsetY = minYOffset;
            break;
    }

    // Координата X
    switch (direction) {
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            newOffsetX = cellRect.x - startPoint.x;
            newOffsetX = newOffsetX > 0 ? minXOffset : newOffsetX;
            break;
        case 'right':
        case 'topRight':
        case 'bottomRight':
            newOffsetX = cellRect.x + cellRect.width - startPoint.x;
            newOffsetX = newOffsetX < 0 ? minXOffset : newOffsetX;
            break;
        case 'top':
        case 'bottom':
            newOffsetX = minXOffset;
            break;
    }

    return {
        x: newOffsetX,
        y: newOffsetY,
    };
}

export function getSelectionPoints(
    startPoint: TPoint,
    direction: TResizerDirection,
    offsetX: number,
    offsetY: number
): {
    start: TPoint;
    end: TPoint;
} {
    let x1: number;
    let y1: number;

    let x2: number;
    let y2: number;

    // Координата Y
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'topRight':
            y1 = startPoint.y - 2;
            y2 = startPoint.y - -1 * offsetY + 2;
            break;
        case 'bottom':
        case 'right':
        case 'bottomLeft':
        case 'bottomRight':
            y1 = startPoint.y + 2;
            y2 = startPoint.y + offsetY - 2;
            break;
    }

    // Координата X
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            x1 = startPoint.x - 2;
            x2 = startPoint.x - -1 * offsetX + 2;
            break;
        case 'bottom':
        case 'right':
        case 'topRight':
        case 'bottomRight':
            x1 = startPoint.x + 2;
            x2 = startPoint.x + offsetX - 2;
            break;
    }

    return {
        start: {
            x: Math.min(x1, x2),
            y: Math.min(y1, y2),
        },
        end: {
            x: Math.max(x1, x2),
            y: Math.max(y1, y2),
        },
    };
}

export function getRectParams(startPoint: TPoint, offsetX: number, offsetY: number) {
    const transform = (startPosition: number, offset: number) => {
        if (offset < 0) {
            return {
                size: offset * -1,
                position: startPosition + offset,
            };
        }
        return {
            size: offset,
            position: startPosition,
        };
    };

    const xParams = transform(startPoint.x, offsetX);
    const yParams = transform(startPoint.y, offsetY);

    return {
        x: xParams.position,
        y: yParams.position,
        width: xParams.size,
        height: yParams.size,
    };
}
