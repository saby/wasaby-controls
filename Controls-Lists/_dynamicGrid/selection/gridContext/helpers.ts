import { TColumnKey, TGridRowSelection, TGridSelection, TPlainSelection } from '../interfaces';

export function resetHasSiblingToDirection(
    selection: TGridRowSelection,
    direction: 'up' | 'down'
): boolean {
    let isChanged = false;

    Object.keys(selection.hasSibling).forEach((key) => {
        if (selection.hasSibling[key][direction]) {
            selection.hasSibling[key][direction] = false;
            isChanged = true;
        }
    });

    return isChanged;
}

function updateSiblingRowSelectionsToDirectionMutable(
    selection: TGridRowSelection,
    siblingSelection: TGridRowSelection,
    currentSelectionKey: TColumnKey,
    direction: 'up' | 'down'
): boolean {
    let isChanged = false;

    const siblingKey = Object.keys(siblingSelection.rowSelection).find(
        (prevKey) =>
            siblingSelection.rowSelection[prevKey].firstColumnKey ===
                selection.rowSelection[currentSelectionKey].firstColumnKey &&
            siblingSelection.rowSelection[prevKey].lastColumnKey ===
                selection.rowSelection[currentSelectionKey].lastColumnKey
    );

    if (typeof siblingKey !== 'undefined') {
        siblingSelection.hasSibling[siblingKey][direction === 'up' ? 'down' : 'up'] = true;
        selection.hasSibling[currentSelectionKey][direction] = true;
        isChanged = true;
    }

    return isChanged;
}

export function updateHasSiblingMutable(
    selection: TGridRowSelection,
    prevSiblingSelection: TGridRowSelection,
    nextSiblingSelection: TGridRowSelection
): [isPrevChanged: boolean, isNextChanged: boolean] {
    Object.keys(selection.rowSelection).forEach((key) => {
        selection.hasSibling[key] = {
            up: false,
            down: false,
        };
    });

    if (!prevSiblingSelection && !nextSiblingSelection) {
        return [false, false];
    }

    let isPrevChanged = false;
    let isNextChanged = false;

    if (prevSiblingSelection) {
        isPrevChanged = resetHasSiblingToDirection(prevSiblingSelection, 'down');
    }

    if (nextSiblingSelection) {
        isNextChanged = resetHasSiblingToDirection(nextSiblingSelection, 'up');
    }

    Object.keys(selection.rowSelection).forEach((currentSelectionKey) => {
        if (prevSiblingSelection) {
            isPrevChanged =
                updateSiblingRowSelectionsToDirectionMutable(
                    selection,
                    prevSiblingSelection,
                    currentSelectionKey as unknown as TColumnKey,
                    'up'
                ) || isPrevChanged;
        }

        if (nextSiblingSelection) {
            isNextChanged =
                updateSiblingRowSelectionsToDirectionMutable(
                    selection,
                    nextSiblingSelection,
                    currentSelectionKey as unknown as TColumnKey,
                    'down'
                ) || isNextChanged;
        }
    });

    if (isPrevChanged) {
        prevSiblingSelection.hasSibling = { ...prevSiblingSelection.hasSibling };
    }

    if (isNextChanged) {
        nextSiblingSelection.hasSibling = { ...nextSiblingSelection.hasSibling };
    }

    return [isPrevChanged, isNextChanged];
}

export function isEqualPlainSelections(
    selection1: TPlainSelection,
    selection2: TPlainSelection
): boolean {
    return compareKeys(Object.keys(selection1), Object.keys(selection2), (a, b) => {
        return compareKeys(selection1[a], selection2[b]);
    });
}

function compareKeys<T>(
    keys1: T[],
    keys2: T[],
    extraComparator?: (key1: T, key2: T) => boolean
): boolean {
    if (keys1.length !== keys2.length) {
        return false;
    } else if (keys1.length === 0) {
        return true;
    } else {
        const sorted1 = keys1.sort();
        const sorted2 = keys2.sort();

        // Ключи есть, их одинаковое количество и они отсортированы.
        for (let i = 0; i < sorted1.length; i++) {
            if (sorted1[i] !== sorted2[i]) {
                return false;
            } else if (extraComparator && !extraComparator(sorted1[i], sorted2[i])) {
                return false;
            }
        }

        return true;
    }
}

export function convertGridSelectionToPlainRecord(selection: TGridSelection): TPlainSelection {
    return Object.keys(selection).reduce((acc, sKey) => {
        return {
            ...acc,
            [sKey]: Object.keys(selection[sKey].rowSelection).reduce(
                (rowSelectedKeys, rsKey) => [
                    ...rowSelectedKeys,
                    ...selection[sKey].rowSelection[rsKey].selectedKeys,
                ],
                []
            ),
        };
    }, {});
}

export function convertPlainSelectionToGridSelection(selection: TPlainSelection): TGridSelection {
    const gridSelection: TGridSelection = {};
    // TODO: Не реализовано, будет делаться тут, если это вообще требуется.
    // Сейчас смена набора выбранных ячеек прикладным разработчиком через слайс не работает.
    return gridSelection;
}
