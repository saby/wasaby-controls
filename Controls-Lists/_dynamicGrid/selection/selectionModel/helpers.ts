/**
 * @kaizen_zone 9377bd5b-f96c-43f4-bb99-324d7bfb4363
 */
import {
    ICellsSelectionModel,
    IRowSelectionModel,
    IGridSelectionModel,
    TSelectionMap,
} from './types';
import { TColumnKey, TColumnKeys, TItemKey } from '../shared/types';

export function createEmptySelection(newKeys: TColumnKey[]): ICellsSelectionModel {
    return {
        rootKey: newKeys[0],
        firstColumnKey: newKeys[0],
        lastColumnKey: newKeys[newKeys.length - 1],
        selectedKeys: newKeys,
        gridColumnStart: undefined,
        gridColumnEnd: undefined,

        isHidden: false,
        hasSiblingUp: false,
        hasSiblingDown: false,
        prevItemKey: undefined,
        nextItemKey: undefined,
    };
}

export function getGridColumnCSSStyle(
    selection: ICellsSelectionModel,
    columns: TColumnKeys
): {
    gridColumnStart: number;
    gridColumnEnd: number;
} {
    let gridColumnStart = columns.indexOf(selection.firstColumnKey);
    let gridColumnEnd = columns.indexOf(selection.lastColumnKey);

    if (gridColumnStart === -1 || gridColumnEnd === -1) {
        if (gridColumnStart === -1 && gridColumnEnd === -1) {
            gridColumnStart = undefined;
            gridColumnEnd = undefined;
        } else if (gridColumnStart === -1) {
            gridColumnStart = 1;
        } else if (gridColumnEnd === -1) {
            gridColumnEnd = -1;
        }
    } else {
        gridColumnStart = gridColumnStart + 1;
        gridColumnEnd = gridColumnEnd + 1 + 1;
    }

    return {
        gridColumnStart,
        gridColumnEnd,
    };
}

export function clone<T>(obj: T): T {
    return JSON.parse(JSON.stringify(obj));
}

export function resetHasSiblingToDirection(
    selection: IRowSelectionModel,
    direction: 'up' | 'down' | 'both'
): boolean {
    let isChanged = false;

    Object.keys(selection).forEach((key) => {
        if (direction === 'up' || direction === 'both') {
            if (selection[key].hasSiblingUp) {
                isChanged = true;
            }
            selection[key].hasSiblingUp = false;
            selection[key].prevItemKey = undefined;
        }
        if (direction === 'down' || direction === 'both') {
            if (selection[key].hasSiblingDown) {
                isChanged = true;
            }
            selection[key].hasSiblingDown = false;
            selection[key].nextItemKey = undefined;
        }
    });

    return isChanged;
}

export function updateSiblingCellsSelection(
    gridSelection: IGridSelectionModel,
    currentItemKey: TItemKey,
    currentRowSelectionKey: TColumnKey,
    siblingItemKey: TItemKey,
    direction: 'up' | 'down'
): boolean {
    let isChanged = false;

    const rowSelection = gridSelection[currentItemKey] as unknown as IRowSelectionModel;
    const siblingRowSelection = gridSelection[siblingItemKey] as unknown as IRowSelectionModel;

    const siblingCellsSelectionRoot = Object.keys(siblingRowSelection).find(
        (prevKey) =>
            siblingRowSelection[prevKey].firstColumnKey ===
                rowSelection[currentRowSelectionKey].firstColumnKey &&
            siblingRowSelection[prevKey].lastColumnKey ===
                rowSelection[currentRowSelectionKey].lastColumnKey
    );

    if (typeof siblingCellsSelectionRoot !== 'undefined') {
        const cellsSelection = rowSelection[currentRowSelectionKey] as ICellsSelectionModel;
        const sCellsSelection = siblingRowSelection[
            siblingCellsSelectionRoot
        ] as ICellsSelectionModel;

        if (direction === 'up') {
            sCellsSelection.hasSiblingDown = true;
            sCellsSelection.nextItemKey = currentItemKey;
            cellsSelection.hasSiblingUp = true;
            cellsSelection.prevItemKey = siblingItemKey;
        } else {
            sCellsSelection.hasSiblingUp = true;
            sCellsSelection.prevItemKey = currentItemKey;
            cellsSelection.hasSiblingDown = true;
            cellsSelection.nextItemKey = siblingItemKey;
        }
        isChanged = true;
    }

    return isChanged;
}

export function isEqualPlainSelections(
    selection1: TSelectionMap,
    selection2: TSelectionMap
): boolean {
    return compareKeys(Object.keys(selection1), Object.keys(selection2), (a, b) => {
        return compareKeys(selection1[a], selection2[b]);
    });
}

export function convertGridSelectionToPlainRecord(selection: IGridSelectionModel): TSelectionMap {
    return Object.keys(selection).reduce((acc, sKey) => {
        return {
            ...acc,
            [sKey]: Object.keys(selection[sKey]).reduce(
                (rowSelectedKeys, rsKey) => [
                    ...rowSelectedKeys,
                    ...selection[sKey][rsKey].selectedKeys,
                ],
                []
            ),
        };
    }, {});
}

function compareKeys<T>(
    keys1: T[],
    keys2: T[],
    extraComparator?: (key1: T, key2: T) => boolean
): boolean {
    if (keys1.length !== keys2.length) {
        return false;
    } else if (keys1.length === 0) {
        return true;
    } else {
        const sorted1 = keys1.sort();
        const sorted2 = keys2.sort();

        // Ключи есть, их одинаковое количество и они отсортированы.
        for (let i = 0; i < sorted1.length; i++) {
            if (sorted1[i] !== sorted2[i]) {
                return false;
            } else if (extraComparator && !extraComparator(sorted1[i], sorted2[i])) {
                return false;
            }
        }

        return true;
    }
}
