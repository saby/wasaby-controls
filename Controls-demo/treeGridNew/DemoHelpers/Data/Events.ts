import { IColumn } from 'Controls/grid';
import { IData } from '../Interface';

export const Events = {
    getColumns: (): IColumn[] => {
        return [
            {
                displayProperty: 'title',
                width: '',
            },
        ];
    },
    getDataBeforeExpanded: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Асинхронная обработка перед разворотом узла',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Асинхронная обработка перед разворотом узла',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataBeforeCollapsed: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Асинхронная обработка перед сворачиванием узла',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Асинхронная обработка перед сворачиванием узла',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataCollapsedItemsChanged: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Этот узел нельзя свернуть',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Этот узел можно свернуть',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataExpandedItemsChanged: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Этот узел нельзя развернуть',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Этот узел можно развернуть',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
};
