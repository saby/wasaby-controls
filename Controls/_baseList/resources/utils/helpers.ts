/**
 * @kaizen_zone b9244594-2ac8-4db9-8c67-cd873d12a1c4
 */
import { Model } from 'Types/entity';
import { CollectionItem } from 'Controls/display';
import { CrudEntityKey } from 'Types/source';

/**
 * TODO нужно выпилить этот метод при переписывании моделей. item.getContents() должен возвращать Record
 *  https://online.sbis.ru/opendoc.html?guid=acd18e5d-3250-4e5d-87ba-96b937d8df13
 * @param item
 */
export function getPlainItemContents(item: CollectionItem<Model>): Model {
    let contents = item.getContents();
    if (item['[Controls/_baseTree/BreadcrumbsItem]'] || item.breadCrumbs) {
        contents = contents[(contents as any).length - 1];
    }
    return contents;
}

export function getValidatedItemKey(key: string): CrudEntityKey {
    // Если в начале строки точка, то нельзя приводить этот ключ к числу, т.к. получим число 0.XXXX
    return isNaN(key) || (typeof key === 'string' && key[0] === '.')
        ? key
        : Number(key);
}

export function getKey(item: Model | string): CrudEntityKey {
    if (item['[Types/_entity/Model]']) {
        return item.getKey();
    }

    return item as string;
}
