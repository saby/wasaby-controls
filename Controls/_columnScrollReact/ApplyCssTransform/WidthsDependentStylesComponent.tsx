import * as React from 'react';
import { IColumnScrollWidths } from '../common/interfaces';
import { ISelectorsState } from '../common/selectors';
import { QA_SELECTORS } from '../common/data-qa';

export interface IWidthsDependentStylesComponentProps
    extends Pick<IColumnScrollWidths, 'startFixedWidth' | 'endFixedWidth' | 'viewPortWidth'> {
    selectors: ISelectorsState;
}

/**
 * @private
 * @pure
 */
export function WidthsDependentStylesComponent(
    props: IWidthsDependentStylesComponentProps
): React.FunctionComponentElement<IWidthsDependentStylesComponentProps> {
    return (
        <style data-qa={QA_SELECTORS.WIDTH_STYLES_TAG}>
            {`.${props.selectors.OFFSET_FOR_START_FIXED_ELEMENT} { left: ${props.startFixedWidth}px; }`}
            {`.${props.selectors.OFFSET_FOR_END_FIXED_ELEMENT} { right: ${props.endFixedWidth}px; } `}
            {`.${props.selectors.STRETCHED_TO_VIEWPORT_ELEMENT} { width: ${props.viewPortWidth}px; }`}
        </style>
    );
}

export const WidthsDependentStylesComponentMemo = React.memo(WidthsDependentStylesComponent);
export default WidthsDependentStylesComponentMemo;
