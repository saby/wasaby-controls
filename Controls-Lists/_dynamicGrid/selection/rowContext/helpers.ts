import { TColumnKey, ISelection, TRowSelection, TColumns } from '../interfaces';
import { datesEqualByQuantum, TQuantumType } from '../../render/utils';

export function getColumnIndexByKey(
    columnKey: TColumnKey,
    columns: TColumns,
    quantum: TQuantumType
): number {
    const date = new Date(columnKey);
    return columns.findIndex((d) => datesEqualByQuantum(date, d, quantum));
}

export function createNewSelection(newKeys: TColumnKey[]): ISelection {
    return {
        rootKey: newKeys[0],
        firstColumnKey: newKeys[0],
        lastColumnKey: newKeys[newKeys.length - 1],
        selectedKeys: newKeys,
        hasBorderTop: true,
        hasBorderBottom: true,
        gridColumnStart: undefined,
        gridColumnEnd: undefined,
    };
}

export function unshiftFromSelectionMutable(
    rowSelection: TRowSelection,
    selection: ISelection,
    columnsToRoot: Record<TColumnKey, TColumnKey>
): ISelection {
    const columnKey = selection.firstColumnKey;
    const rightSelection: ISelection = createNewSelection(selection.selectedKeys.slice(1));

    rightSelection.selectedKeys.forEach((k) => {
        columnsToRoot[k] = rightSelection.rootKey;
    });

    delete rowSelection[columnsToRoot[columnKey]];
    delete columnsToRoot[columnKey];

    rowSelection[rightSelection.rootKey] = rightSelection;

    return rightSelection;
}

export function popFromSelectionMutable(
    rowSelection: TRowSelection,
    selection: ISelection,
    columnsToRoot: Record<TColumnKey, TColumnKey>
): ISelection {
    const columnKey = selection.lastColumnKey;
    const leftSelection: ISelection = createNewSelection(
        selection.selectedKeys.slice(0, selection.selectedKeys.length - 1)
    );

    leftSelection.selectedKeys.forEach((k) => {
        columnsToRoot[k] = leftSelection.rootKey;
    });

    delete rowSelection[columnsToRoot[columnKey]];
    delete columnsToRoot[columnKey];
    rowSelection[leftSelection.rootKey] = leftSelection;

    return leftSelection;
}

export function splitSelectionMutable(
    rowSelection: TRowSelection,
    selection: ISelection,
    columnsToRoot: Record<TColumnKey, TColumnKey>,
    columnKey: TColumnKey
): [leftSelection: ISelection, rightSelection: ISelection] {
    const leftSelection = createNewSelection(
        selection.selectedKeys.slice(0, selection.selectedKeys.indexOf(columnKey))
    );

    leftSelection.selectedKeys.forEach((k) => {
        columnsToRoot[k] = leftSelection.rootKey;
    });

    const rightSelection = createNewSelection(
        selection.selectedKeys.slice(selection.selectedKeys.indexOf(columnKey) + 1)
    );

    rightSelection.selectedKeys.forEach((k) => {
        columnsToRoot[k] = rightSelection.rootKey;
    });

    delete rowSelection[columnsToRoot[columnKey]];
    delete columnsToRoot[columnKey];

    rowSelection[leftSelection.rootKey] = leftSelection;
    rowSelection[rightSelection.rootKey] = rightSelection;

    return [leftSelection, rightSelection];
}

export function deleteSelectionMutable(
    rowSelection: TRowSelection,
    selection: ISelection,
    columnsToRoot: Record<TColumnKey, TColumnKey>
): void {
    const columnKey = selection.rootKey;
    delete rowSelection[columnsToRoot[columnKey]];
    delete columnsToRoot[columnKey];
}

export function handleAddToSelectionMutable(
    rowSelection: TRowSelection,
    columnsToRoot: Record<TColumnKey, TColumnKey>,
    columnKey: TColumnKey,
    prevColumnKey: TColumnKey,
    nextColumnKey: TColumnKey
): ISelection {
    let joinSelection: ISelection;

    if (columnsToRoot[prevColumnKey]) {
        joinSelection = rowSelection[columnsToRoot[prevColumnKey]];

        joinSelection.selectedKeys.push(columnKey);
        joinSelection.lastColumnKey = columnKey;
        columnsToRoot[columnKey] = columnsToRoot[prevColumnKey];
    }

    if (columnsToRoot[nextColumnKey]) {
        const nextSelection = rowSelection[columnsToRoot[nextColumnKey]];
        const nextSelectionRoot = columnsToRoot[nextColumnKey];

        if (joinSelection) {
            joinSelection.selectedKeys.push(...nextSelection.selectedKeys);
            joinSelection.lastColumnKey = nextSelection.lastColumnKey;
            nextSelection.selectedKeys.forEach((key) => {
                columnsToRoot[key] = joinSelection.rootKey;
            });
            delete rowSelection[nextSelectionRoot];
        } else {
            joinSelection = nextSelection;

            joinSelection.selectedKeys.unshift(columnKey);
            joinSelection.firstColumnKey = columnKey;
            columnsToRoot[columnKey] = columnsToRoot[nextColumnKey];
        }
    }

    if (!joinSelection) {
        joinSelection = createNewSelection([columnKey]);
        rowSelection[columnKey] = joinSelection;
        columnsToRoot[columnKey] = columnKey;
    }

    return joinSelection;
}

export function getGridColumnCSSStyle(selection: ISelection, columns: TColumns, quantum: TQuantumType): {
    gridColumnStart: number;
    gridColumnEnd: number;
} {
    let gridColumnStart = getColumnIndexByKey(selection.firstColumnKey, columns, quantum);
    let gridColumnEnd = getColumnIndexByKey(selection.lastColumnKey, columns, quantum);

    if (gridColumnStart === -1 || gridColumnEnd === -1) {
        if (gridColumnStart === -1 && gridColumnEnd === -1) {
            gridColumnStart = undefined;
            gridColumnEnd = undefined;
        } else if (gridColumnStart === -1) {
            gridColumnStart = 1;
        } else if (gridColumnEnd === -1) {
            gridColumnEnd = -1;
        }
    } else {
        gridColumnStart = gridColumnStart + 1;
        gridColumnEnd = gridColumnEnd + 1 + 1;
    }

    return {gridColumnStart, gridColumnEnd};
}
