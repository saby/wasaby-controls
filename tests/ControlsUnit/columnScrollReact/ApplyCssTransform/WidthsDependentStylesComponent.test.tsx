/**
 * @jest-environment jsdom
 */
import { unmountComponentAtNode } from 'react-dom';
import { WasabyEvents } from 'UICore/Events';

import WidthsDependentStylesComponent from 'Controls/_columnScrollReact/ApplyCssTransform/WidthsDependentStylesComponent';
import { SELECTORS } from '../SelectorsForTests';
import { renderStyleInnerPureComponent } from '../forTestsOnly/helpers';

describe('ControlsUnit/columnScrollReact/ApplyCssTransform/WidthsDependentStylesComponent', () => {
    let container = null;

    beforeEach(() => {
        container = document.createElement('div');
        WasabyEvents.initInstance(container);
        document.body.appendChild(container);
    });

    afterEach(() => {
        unmountComponentAtNode(container);
        WasabyEvents.destroyInstance(container);
        container.remove();
        container = null;
    });

    it('should render width styles.', () => {
        const { getStyleTag } = renderStyleInnerPureComponent(
            <WidthsDependentStylesComponent
                selectors={SELECTORS}
                fixedWidth={702}
            />,
            { container }
        );
        expect(getStyleTag()).toMatchSnapshot();
    });
});
