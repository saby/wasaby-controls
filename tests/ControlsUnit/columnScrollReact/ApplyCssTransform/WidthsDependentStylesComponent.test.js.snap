// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlsUnit/columnScrollReact/ApplyCssTransform/WidthsDependentStylesComponent should render width styles. 1`] = `
<style
  data-qa="ColumnScrollReact__width-styles"
>
  .js-controls-ColumnScrollReact_offsetForFixed_guid123 { left: 702px; }
  .js-controls-ColumnScrollReact_offsetForEndFixed_guid123 { right: 0px; } 
  .js-controls-ColumnScrollReact_stretchedToViewPortElement_guid123 { width: 1200px; }
</style>
`;
